cs_tools/sync/postgres/syncer.py

```diff
@@ -0,0 +1,119 @@
+    log_level: Literal["debug", "info", "warning"] = "warning"
+    secret: Optional[str] = None
+    private_key_path: Optional[pydantic.FilePath] = None
+        
+    @pydantic.field_validator("secret", mode="before")
+    def ensure_auth_secret_given(cls, value: Any, info: pydantic.ValidationInfo) -> Any:
+        if info.data.get("authentication") == "sso" or value is not None:
+            return value
+
+        if info.data.get("authentication") == "basic":
+            raise PydanticCustomError("missing", "Field required, you must provide a password", {"secret": value})
+
+        if info.data.get("authentication") == "oauth":
+            raise PydanticCustomError("missing", "Field required, you must provide an oauth token", {"secret": value})
+    
+        
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        logging.getLogger("postgres").setLevel(self.log_level.upper())
+        
+        self._engine = sa.create_engine(self.make_url())
+        self.Session = sessionmaker(bind=self.engine)
+        self.metadata = sqlmodel.MetaData(schema=self.schema_)
+
+    def __repr__(self) -> str:
+        database = self.dbname
+        username = self.user
+        host = self.host
+        port = self.port
+        return f"<PostgresSyncer DATABASE='{database}', USER='{username}', HOST='{host}', PORT='{port}'>"
+    
+    def make_url(self):
+        """Create a connection string for the Postgres JDBC driver."""
+        dialect = "postgresql"
+        driver = "psycopg2"
+        connection_string = f"{dialect}+{driver}://{self.user}:{self.password}@{self.host}/{self.dbname}"
+        return connection_string
+    
+    # MANDATORY PROTOCOL MEMBERS
+    
+    @staticmethod
+    def output_to_list(rows):
+        output_list = []
+        for row in rows:
+            values = [str(value) for value in row]
+            output_list.append(values)
+        return output_list
+    
+    def load(self, tablename: str) -> TableRows:
+        """SELECT rows from Postgresql SQL."""
+        table = self.metadata.tables[f"{self.schema_}.{tablename}"]
+        rows = self.session.execute(table.select())
+        result = self.output_to_list(rows) 
+        return result
+
+    def dump(self, tablename: str, *, data: TableRows) -> None:
+        """INSERT rows into Postgres SQL."""
+        if not data:
+            log.warning(f"no data to write to syncer {self}")
+            return
+
+        table = self.metadata.tables[f"{self.schema_}.{tablename}"]
+
+        if self.load_strategy == "APPEND":
+            sync_utils.batched(table.insert().values, session=self.session, data=data)
+
+        if self.load_strategy == "TRUNCATE":
+            self.session.execute(table.delete())
+            sync_utils.batched(table.insert().values, session=self.session, data=data)
+
+        if self.load_strategy == "UPSERT":
+            sync_utils.generic_upsert(table, session=self.session, data=data)
```

Comment:
Nice job here! No changes required. Thank you for looking at other Syncers and finding the patterns we're using in them.

Couple of questions I have.. maybe you found them in your research.

- Is there a maximum number of query/statement parameters in a single pg statements? This would need to go as an argument to `sync_utils.batched` to avoid an `sqlalchemy.exc.OperationalError`.

- Do you know if pg supports the `PUT -> COPY INTO/MERGE` data load pattern? This is something that greatly increased the throughput on Snowflake syncer, so I want to explore this wherever we can.

If you don't know the answer, we can leave it for future improvements.

Reply:
By default, max_prepared_transactions is set to 0, meaning that prepared transactions are disabled. However, we can specify the batch size as an argument to limit the number of parameters in each batch. I have modifed the load method to include batch size where in this method, it can handle large datasets.

```
def load(self, tablename: str, batch_size: int = 1000) -> TableRows:
        """SELECT rows from Databricks"""
        table = self.metadata.tables[f"{self.schema_}.{tablename}"]
        rows = self.session.execute(table.select())
        result = []
        for batch in self.batched(rows, batch_size):
            result.extend([row.model_dump() for row in batch])
        return result

    def batched(self, iterable, n):
        """Yield successive n-sized chunks from iterable."""
        for i in range(0, len(iterable), n):
            yield iterable[i:i + n]
```

Reply:
As part of my research, I don't think Postgres supports to stage our data and copy into our dataset without any third party interference. Although we can use alternative ways to do this, but need to look at this as part of future improvements. 

I have added a comment to the same in the script.